Make clickable, mouseClickable, toggleable request focus onClick when the InputMode is Keyboard

eymar opened the pull request:
Changes in this PR:
- replace `focusableInNonTouchMode` usages by simple `focusable` (this change assumes that the purpose of the former was to simply prevent Focus indication in Touch mode. We want to focus on `focusable` items without any indications in Touch mode. This ensures that when users start navigating with Tab, the focus will move from relevant component to its relevant neighbours)
- add `focusRequester` in `clickable` and `toggleable` and request focus when Tap event occurs 
- make `Modifier.indication` skip FocusInteraction.Focus when `LocalInputModeManager` is in Touch mode
- make SkiaBasedOwner distinguish Touch and Keyboard input modes

eymar committed:
Make clickable, mouseClickable, toggleable request focus onClick when the InputMode is Keyboard

igordmn commented on compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/Clickable.kt:
```diff
@@ -138,6 +153,7 @@ fun Modifier.clickable(
         val delayPressInteraction = rememberUpdatedState {
             isClickableInScrollableContainer.value || isRootInScrollableContainer()
         }
+        val (focusRequester, focusRequesterModifier) = focusRequesterForKeyboardMode()
```
Thanks! Code looks good.

1.
I am worried about indication. Especially for material, as it looks bulky:

https://user-images.githubusercontent.com/5963351/179999326-b6aeaffa-7f59-4b45-b6f6-9f06bbe11359.mov

We probably should introduce some state, similar to touchMode in Android. Something like isKeyboardFocusing, which we set to true when press Tab, and to false after any mouse down. Or maybe even use `touchMode` itself? I am not sure. We should hide focusable indication, when we don't interact via focus keys.

2.
Also, it seems we have a bug with focusable indication - hover indication overrides it:

https://user-images.githubusercontent.com/5963351/180000863-57425736-d879-4841-9981-7e249ea8879d.mov

It is a separate bug, and can be fixed separately.

3.
Also, we probably should add this feature for Slider (it isn't clickable, but focusable, and we can interact with it)

igordmn replied:
> We should hide focusable indication, when we don't interact via focus keys.

And we don't have to introduce the new interaction type (like HiddenFocused), we can check this state inside the indication implementation.

eymar replied:
> I am worried about indication. Especially for material, as it looks bulky:

The problem with indication has been solved in the last commit by making Modifier.indication prevent drawing the Focus Indication. 

> Also, we probably should add this feature for Slider (it isn't clickable, but focusable, and we can interact with it)

wip

eymar replied:
@igordmn the last commit has a change with Slider: now when user ends the gesture (tap. drag) on a Slider, the focus will be requested.

igordmn replied:
Looks good to me, thanks!

eymar committed:
Make Modifier.indication prevent drawing the Focus indication when inputMode is Touch

Also, replace `focusableInTouchMode` usages by `focusable`, since it was used only to prevent the Focus Indication drawing. But now we want to make them focusable in all modes, but without the Focus indication in Touch mode.

eymar committed:
Add focusRequester to Slider implementation and request focus in gestureEndAction
